File: frontend/src/types/portfolio.ts
export type TransactionType = 'BUY' | 'SELL';

/** `fallback` means CoinGecko was unreachable and the average buy price was used instead. */
export type PriceSource = 'live' | 'fallback';

export interface PortfolioAsset {
  asset_id: number;
  coin_id: string;
  symbol: string;
  name: string;
  image_url: string | null;
  quantity: number;
  avg_buy_price: number;
  current_price: number;
  price_change_24h: number | null;
  price_source: PriceSource;
  value: number;
  invested: number;
  profit_loss: number;
  profit_loss_pct: number;
  allocation_pct: number;
}

export interface Portfolio {
  cash_balance: number;
  initial_balance: number;
  invested_amount: number;
  holdings_value: number;
  total_value: number;
  profit_loss: number;
  performance_pct: number;
  assets: PortfolioAsset[];
  updated_at: string;
}

export interface Transaction {
  id: number;
  coin_id: string;
  symbol: string;
  name: string;
  image_url: string | null;
  type: TransactionType;
  quantity: number;
  price: number;
  total: number;
  created_at: string;
}

export interface PortfolioSnapshot {
  total_value: number;
  profit_loss: number;
  created_at: string;
}

export interface SnapshotsResponse {
  days: number;
  initial_balance: number;
  points: PortfolioSnapshot[];
}

export interface TradeRequest {
  coin_id: string;
  quantity: number;
}

export interface TradeResponse {
  message: string;
  transaction: Transaction;
  portfolio: Portfolio;
}

File: backend/app/schemas/portfolio.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict

TransactionTypeLiteral = Literal["BUY", "SELL"]
PriceSource = Literal["live", "fallback"]


class PortfolioAsset(BaseModel):
    asset_id: int
    coin_id: str
    symbol: str
    name: str
    image_url: str | None = None
    quantity: float
    avg_buy_price: float
    current_price: float
    price_change_24h: float | None = None
    price_source: PriceSource = "live"
    value: float
    invested: float
    profit_loss: float
    profit_loss_pct: float
    allocation_pct: float


class PortfolioSummary(BaseModel):
    cash_balance: float
    initial_balance: float
    invested_amount: float
    holdings_value: float
    total_value: float
    profit_loss: float
    performance_pct: float
    assets: list[PortfolioAsset]
    updated_at: datetime


class TransactionOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    coin_id: str
    symbol: str
    name: str
    image_url: str | None = None
    type: TransactionTypeLiteral
    quantity: float
    price: float
    total: float
    created_at: datetime


class SnapshotOut(BaseModel):
    total_value: float
    profit_loss: float
    created_at: datetime


class SnapshotsResponse(BaseModel):
    days: int
    initial_balance: float
    points: list[SnapshotOut]
